
:GeneralizeCurriedFunctions


def f n = n 0
//│ f: (0 -> 'a & 'b) -> 'a
//│  = [Function: f]

:e
f {}
//│ ╔══[ERROR] Type mismatch in application:
//│ ║  l.10: 	f {}
//│ ║        	^^^^
//│ ╟── record literal of type `anything` is not a function
//│ ║  l.10: 	f {}
//│ ║        	  ^^
//│ ╟── Note: constraint arises from application:
//│ ║  l.5: 	def f n = n 0
//│ ║       	          ^^^
//│ ╟── from reference:
//│ ║  l.5: 	def f n = n 0
//│ ╙──     	          ^
//│ res: error
//│ Runtime error:
//│   TypeError: n is not a function

// :d
:e
f {} {}
//│ ╔══[ERROR] Type mismatch in application:
//│ ║  l.29: 	f {} {}
//│ ║        	^^^^
//│ ╟── record literal of type `anything` is not a function
//│ ║  l.29: 	f {} {}
//│ ║        	  ^^
//│ ╟── Note: constraint arises from application:
//│ ║  l.5: 	def f n = n 0
//│ ║       	          ^^^
//│ ╟── from reference:
//│ ║  l.5: 	def f n = n 0
//│ ╙──     	          ^
//│ res: error
//│ Runtime error:
//│   TypeError: n is not a function


def f n m = n 0
//│ f: 'a -> (forall 'a, 'b, 'c. ('c -> 'b
//│   where
//│     'a <: 0 -> 'b))
//│  = [Function: f1]

// :e // Error delayed by inconsistent constrained types
f {}
//│ res: 'a -> 'b
//│   where
//│     'c <: 0 -> 'b
//│    = [Function (anonymous)]

:e
f {} {}
//│ ╔══[ERROR] Type mismatch in application:
//│ ║  l.61: 	f {} {}
//│ ║        	^^^^^^^
//│ ╟── record literal of type `anything` is not a function
//│ ║  l.61: 	f {} {}
//│ ║        	  ^^
//│ ╟── Note: constraint arises from application:
//│ ║  l.47: 	def f n m = n 0
//│ ╙──      	            ^^^
//│ res: error
//│ Runtime error:
//│   TypeError: n is not a function



def test extr =
  let f n m = n 0
  in let _ = extr f
  in f 0
//│ test: ((forall 'a. 'a -> (forall 'a, 'b, 'c. ('c -> 'b
//│   where
//│     'a <: 0 -> 'b))) -> 'd & 'e) -> (forall 'f, 'g, 'h, 'i. ('h -> 'g | 'i
//│   where
//│     'f <: 0 -> 'g))
//│     = [Function: test]

:e
test 1
//│ ╔══[ERROR] Type mismatch in application:
//│ ║  l.89: 	test 1
//│ ║        	^^^^^^
//│ ╟── integer literal of type `1` is not a function
//│ ║  l.89: 	test 1
//│ ║        	     ^
//│ ╟── Note: constraint arises from application:
//│ ║  l.79: 	  in let _ = extr f
//│ ╙──      	             ^^^^^^
//│ res: error
//│ Runtime error:
//│   TypeError: extr is not a function

// * Note: this extrudes a Constrainedtype referring to a higher-level variable (x)
def test extr x =
  let f n m = x (n 0)
  in let _ = extr f
  in f 0
//│ test: ((forall 'a. 'a -> (forall 'a, 'b, 'c, 'd, 'e. ('e -> 'd
//│   where
//│     'b <: 'c -> 'd
//│     'a <: 0 -> 'c))) -> 'f & 'g) -> (forall 'h. ('b & 'h) -> (forall 'h, 'i, 'j, 'k, 'l, 'm. ('l -> 'j | 'm
//│   where
//│     'h <: 'i -> 'j
//│     'k <: 0 -> 'i)))
//│     = [Function: test1]

:e
test 0
//│ ╔══[ERROR] Type mismatch in application:
//│ ║  l.118: 	test 0
//│ ║         	^^^^^^
//│ ╟── integer literal of type `0` is not a function
//│ ║  l.118: 	test 0
//│ ║         	     ^
//│ ╟── Note: constraint arises from application:
//│ ║  l.106: 	  in let _ = extr f
//│ ╙──       	             ^^^^^^
//│ res: error
//│    = [Function (anonymous)]

def test extr x =
  let f n m = x (n 0)
  in f 0
//│ test: 'a -> (forall 'b. 'b -> (forall 'b, 'c, 'd, 'e, 'f, 'g. ('f -> 'd | 'g
//│   where
//│     'b <: 'c -> 'd
//│     'e <: 0 -> 'c)))
//│     = [Function: test2]

// Error delayed by inconsistent constrained types
test 0 1
//│ res: 'a -> 'b
//│   where
//│     'c <: 'd -> 'b
//│     'e <: 0 -> 'd
//│    = [Function (anonymous)]

:e
test 0 1 2
//│ ╔══[ERROR] Type mismatch in application:
//│ ║  l.149: 	test 0 1 2
//│ ║         	^^^^^^^^^^
//│ ╟── integer literal of type `1` is not a function
//│ ║  l.149: 	test 0 1 2
//│ ║         	       ^
//│ ╟── Note: constraint arises from application:
//│ ║  l.132: 	  let f n m = x (n 0)
//│ ╙──       	              ^^^^^^^
//│ res: error
//│ Runtime error:
//│   TypeError: n is not a function

def test extr x =
  let f n m = (n x 0)
  in f 0
//│ test: 'a -> (forall 'b. 'b -> (forall 'b, 'c, 'd, 'e, 'f, 'g. ('f -> 'd | 'g
//│   where
//│     'c <: 'b -> (0 -> 'd & 'e))))
//│     = [Function: test3]

// Error delayed by inconsistent constrained types
test 0 1
//│ res: 'a -> 'b
//│   where
//│     'c <: 1 -> (0 -> 'b & 'd)
//│    = [Function (anonymous)]

:e
test 0 1 2
//│ ╔══[ERROR] Type mismatch in application:
//│ ║  l.179: 	test 0 1 2
//│ ║         	^^^^^^^^^^
//│ ╟── integer literal of type `0` is not a function
//│ ║  l.165: 	  in f 0
//│ ║         	       ^
//│ ╟── Note: constraint arises from application:
//│ ║  l.164: 	  let f n m = (n x 0)
//│ ╙──       	               ^^^
//│ res: error
//│ Runtime error:
//│   TypeError: n is not a function


